sub/internal/service/sub/integration_test.go

```diff
@@ -0,0 +1,169 @@
+//go:build integration
+
+package sub
+
+import (
+	"context"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/stretchr/testify/require"
+
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-hrvadl/sub/internal/service/validator"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-hrvadl/sub/internal/storage/platform/db"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-hrvadl/sub/internal/storage/subscriber"
+)
+
+const testDSNEnvKey = "SUB_TEST_DSN"
+
+func TestMain(t *testing.M) {
+	code := t.Run()
+	dsn := os.Getenv(testDSNEnvKey)
+
+	db, err := db.NewConn(dsn)
+	if err != nil {
+		panic("failed to connect to test db")
+	}
+
+	if _, err := db.Exec("DELETE FROM subscribers"); err != nil {
+		panic("failed to cleanup")
+	}
+
+	os.Exit(code)
+}
+
+func TestServiceSubscribeInt(t *testing.T) {
+	type args struct {
+		ctx  context.Context
+		mail string
+	}
+	testCases := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "Should subscribe correctly",
+			args: args{
+				ctx:  context.Background(),
+				mail: "test@mail.com",
+			},
+			wantErr: false,
+		},
+		{
+			name: "Should not subscribe correctly when it takes too long",
+			args: args{
+				ctx:  newImmediateCtx(),
+				mail: "test1111@mail.com",
+			},
+			wantErr: true,
+		},
+		{
+			name: "Should not subscribe when email is empty",
+			args: args{
+				ctx:  context.Background(),
+				mail: "",
+			},
+			wantErr: true,
+		},
+		{
+			name: "Should not subscribe when email is incorrect",
+			args: args{
+				ctx:  context.Background(),
+				mail: "tetmail.com",
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			dsn := os.Getenv(testDSNEnvKey)
+			require.NotZero(t, dsn, "test DSN can not be empty")
+			db, err := db.NewConn(dsn)
+			require.NoError(t, err, "Failed to connect to db")
+
+			rs := subscriber.NewRepo(db)
+			v := validator.NewStdlib()
+			s := NewService(rs, v)
+
+			id, err := s.Subscribe(tt.args.ctx, tt.args.mail)
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+
+			t.Cleanup(func() {
```

Comment:
Isn't it better to move cleanup to the start of the test function to ensure its exection?

Reply:
@Hukyl 
the cleanup function is dependent on the ID of the subscriber I [created](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-hrvadl/pull/3/files/e8c69036b27fa83bee2d501103a00baa45f4d8dd#diff-4b11bac90d592dacc79d03fc71bc800461da4904a5d8308c084674dc170c31a9R92)

so, unfortunately I cannot move it anywhere below